src/workerd/api/sql.c++

```diff
@@ -115,24 +154,85 @@ double SqlStorage::Cursor::getRowsWritten() {
   }
 }
 
+SqlStorage::Cursor::RowIterator::Next SqlStorage::Cursor::next(jsg::Lock& js) {
+  KJ_IF_SOME(s, state) {
+    cachedColumnNames.ensureInitialized(js, s->query);
+  }
+
+  auto self = JSG_THIS;
+  auto maybeRow = rowIteratorNext(js, self);
+  bool done = maybeRow == kj::none;
+  return {
+    .done = done,
+    .value = kj::mv(maybeRow),
```

Comment:
I know why you've done this (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_iterator_protocol) but it doesn't actually make this object iterable unless we make `[Symbol.iterator]()` return `this` (which we could do?)

Maybe that's exactly what we want, and then `.raw()` returns a copy of the Cursor but with a different implementation of `next()` that returns arrays. That makes things as JS-standardy as I think we can get?

Reply:
I actually came here to write that we should make `.next()` return `T | null` and then also add a `.done` property:

```js
const row = sql.exec('...').next()
// instead of
const { value: row } = sql.exec('...').next()
```

And `.done` means could check for 0 results without actually calling `.next()`:

```js
const cursor = sql.exec('...')
if (!cursor.done) throw new Error("Wasn't expecting results!!")
```

But both of those seem like marginal benefits, at the cost of diverging from the Iterator standard. Anyway, thought I'd note them here for completeness.

Reply:
@justin-mp you mentioned being a bit confused re: Iterators and Iterables, and I admit I needed to refresh my memory, so I just wrote out a couple of examples to confirm the way things work re: `Symbol.iterator` and `next()`:

```ts
class CustomNext {
  private iterator: Iterator<any>
  public done: boolean = false

  constructor(cursor: SqlStorageCursor) {
    this.iterator = cursor[Symbol.iterator]()
  }

  // Can return anything we want, will never be called
  // by a for-of loop or an array splat because we don't
  // implement [Symbol.iterator]()
  next() {
    const { value, done } = this.iterator.next()
    this.done = Boolean(done)
    return value
  }
}

class CustomIterator {
  private iterator: Iterator<any>

  constructor(cursor: SqlStorageCursor) {
    this.iterator = cursor[Symbol.iterator]()
  }

  // This is what is called when one of these objects is
  // passed to a for-of loop or [...splat] syntax. We must
  // return an object that implements next() => { value, done }
  [Symbol.iterator]() {
    return this
  }

  // Must follow the Iterator protocol
  next() {
    const { value, done } = this.iterator.next()
    return { value, done }
  }
}
```

Anyway, I am now fully convinced we should follow the `CustomIterator` example above (follow the spec) vs the `CustomNext` example (yolo)